src/org/intellij/erlang/compilation/ErlangPrepareDependenciesCompileTask.java

```diff
@@ -128,161 +128,141 @@ private static ErlangModuleBuildOrderDescriptor getModuleBuildOrderInner(Module
+  private static List<ErlangFileDescriptor> getTopologicallySortedErlangFilesPaths(@NotNull Collection<ErlangFile> erlangModules,
+                                                                                   @NotNull Collection<ErlangFile> erlangHeaders,
+                                                                                   @NotNull List<String> globalParseTransforms) throws CyclicDependencyFoundException {
+    final ErlangFilesDependencyGraph semiGraph = new ErlangFilesDependencyGraph(erlangModules, erlangHeaders, globalParseTransforms);
+    DFSTBuilder<String> builder = new DFSTBuilder<String>(GraphGenerator.create(semiGraph));
+    builder.buildDFST();
+    if (!builder.isAcyclic()) {
+      throw new CyclicDependencyFoundException();
+    }
+    return ContainerUtil.map(builder.getSortedNodes(), new Function<String, ErlangFileDescriptor>() {
+      @Override
+      public ErlangFileDescriptor fun(String filePath) {
+        return new ErlangFileDescriptor(filePath, semiGraph.getDependencies(filePath));
+      }
+    });
   }
 
-  private static class ErlangModulesSorter {
-    private final Collection<ErlangFile> myModules;
-    private final List<String> myGlobalParseTransforms;
+  private static class ErlangFilesDependencyGraph implements GraphGenerator.SemiGraph<String> {
+    private final HashMap<String, String> myNamesToPathsMap = ContainerUtil.newHashMap();
+    private final HashMap<String, Collection<String>> myPathsToDependenciesMap = ContainerUtil.newHashMap();
 
-    private ErlangModulesSorter(Collection<ErlangFile> modules, List<String> globalParseTransforms) {
-      myModules = modules;
-      myGlobalParseTransforms = globalParseTransforms;
+    public ErlangFilesDependencyGraph(@NotNull Collection<ErlangFile> erlangModules,
+                                      @NotNull Collection<ErlangFile> erlangHeaders,
+                                      @NotNull List<String> globalParseTransforms) {
+      buildNamesMap(erlangModules);
+      buildForHeaders(erlangHeaders);
+      buildForModules(erlangModules, globalParseTransforms);
     }
 
-    public static List<ErlangFileDescriptor> sort(Collection<ErlangFile> erlangModules,
-                                                  List<String> globalParseTransforms) throws CyclicDependencyFoundException {
-      ErlangModulesSorter sorter = new ErlangModulesSorter(erlangModules, globalParseTransforms);
-      return sorter.getSortedModules();
+    private void buildForModules(@NotNull Collection<ErlangFile> erlangModules,
+                                 @NotNull List<String> globalParseTransforms) {
+      List<String> globalPtNodes = getModulePaths(globalParseTransforms);
+      for (ErlangFile erlangModule : erlangModules) {
+        Set<String> dependencies = ContainerUtil.newHashSet();
+        addDeclaredDependencies(erlangModule, dependencies);
+        dependencies.addAll(getPathsFromNames(globalPtNodes));
```

Comment:
```globalPtNodes``` contains paths, which are used as names in this line.

Also ```getModulePaths``` and ```getPathsFromNames``` are the same.